arch/x86/pagetables.c

```diff
@@ -246,28 +249,185 @@ void *vmap_user(void *va, mfn_t mfn, unsigned int order,
+        return -EINVAL;
+    }
+
+    tab = init_map_mfn(cr3_ptr->mfn);
+#if defined(__x86_64__)
+    pml4_t *l4e = l4_table_entry((pml4_t *) tab, va);
+    if (mfn_invalid(l4e->mfn) || !l4e->P) {
+        spin_unlock(&vmap_lock);
+        return -ENOENT;
+    }
+
+    tab = init_map_mfn(l4e->mfn);
+#endif
+    pdpe_t *l3e = l3_table_entry((pdpe_t *) tab, va);
+    if (mfn_invalid(l3e->mfn) || !l3e->P) {
+        spin_unlock(&vmap_lock);
+        return -ENOENT;
+    }
+
+    if (l3e->PS) {
+        set_pgentry(&l3e->entry, MFN_INVALID, PT_NO_FLAGS);
+        spin_unlock(&vmap_lock);
+        return 0;
+    }
+
+    tab = init_map_mfn(l3e->mfn);
+    pde_t *l2e = l2_table_entry((pde_t *) tab, va);
+    if (mfn_invalid(l2e->mfn) || !l2e->P) {
+        spin_unlock(&vmap_lock);
+        return -ENOENT;
+    }
+
+    if (l2e->PS) {
+        set_pgentry(&l2e->entry, MFN_INVALID, PT_NO_FLAGS);
+        spin_unlock(&vmap_lock);
+        return 0;
+    }
+
+    tab = init_map_mfn(l2e->mfn);
+    pte_t *l1e = l1_table_entry((pte_t *) tab, va);
+    if (mfn_invalid(l1e->mfn) || !l1e->P) {
+        spin_unlock(&vmap_lock);
+        return -ENOENT;
+    }
+
+    set_pgentry(&l1e->entry, MFN_INVALID, PT_NO_FLAGS);
+    spin_unlock(&vmap_lock);
+    return 0;
+}
+
+int vunmap_kern(void *va) {
+    return _vunmap(&cr3, va);
+}
+
+int vunmap_user(void *va) {
+    return _vunmap(&user_cr3, va);
+}
+
+#define MMAP_RANGE_SIZE(size, mfn, flags, mmap_flags)                                    \
+    do {                                                                                 \
+        if (likely((mmap_flags) &MMAP_KERNEL))                                           \
+            kmap_##size((mfn), (flags));                                                 \
+        if ((mmap_flags) &MMAP_IDENT)                                                    \
+            vmap_##size(mfn_to_virt((mfn)), (mfn), (flags));                             \
+        if ((mmap_flags) &MMAP_USER)                                                     \
+            vmap_user_##size(mfn_to_virt_user((mfn)), (mfn), (flags | _PAGE_USER));      \
+        if ((mmap_flags) &MMAP_USER_KERNEL)                                              \
+            vmap_##size(mfn_to_virt_user((mfn)), (mfn), (flags & ~_PAGE_USER));          \
+        if ((mmap_flags) &MMAP_KERNEL_MAP)                                               \
+            vmap_##size(mfn_to_virt_map((mfn)), (mfn), (flags));                         \
+    } while (0)
+
+#define MMAP_RANGE_4K(mfn, flags, mmap_flags) MMAP_RANGE_SIZE(4k, mfn, flags, mmap_flags)
+#define MMAP_RANGE_2M(mfn, flags, mmap_flags) MMAP_RANGE_SIZE(2m, mfn, flags, mmap_flags)
+#define MMAP_RANGE_1G(mfn, flags, mmap_flags) MMAP_RANGE_SIZE(1g, mfn, flags, mmap_flags)
+
+void *mmap_range(paddr_t paddr, size_t size, unsigned long flags,
+                 mmap_flags_t mmap_flags) {
+    paddr_t cur = paddr;
+    paddr_t end = cur + size;
```

Comment:
As the *map functions require to be passed a properly aligned address, should the code enforce this too? Or is this part of the contract, i.e. the callers responsibility?

Either way, we should check (and possibly handle) non-page-aligned `cur`.

Reply:
I am not sure I understand fully. The *map functions take two parameters:
1. where to map (the `va`)
2. what to map (the `mfn`)
They have to check if the `va` is properly aligned for the mapping according to the order of the `mfn`.

Here, the `mmap_range()` takes a physical address and turns it into a `mfn`, so it will automatically take care of the non-page-aligned addresses. It maps it automatically to the aligned addresses using various `mfn_to_virt_*()` helpers.

Reply:
Ahh, right. The conversion to a page frame number for the *map() calls implicitly aligns `cur` and the conversion at the end using `paddr_to_virt_map()` will preserve any possible offset.

Sorry for the noise, somehow overlooked the `paddr_to_mfn()` in the macros maze :D